controller/src/main/java/io/pravega/controller/task/Stream/TxnSweeper.java

```diff
@@ -0,0 +1,154 @@
+    private static class Result {
+        private final TxnResource txnResource;
+        private final Object value;
+        private final Throwable error;
+    }
+
+    public TxnSweeper(final StreamMetadataStore streamMetadataStore,
+                      final StreamTransactionMetadataTasks transactionMetadataTasks,
+                      final long maxTxnTimeoutMillis,
+                      final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetadataStore, "streamMetadataStore");
+        Preconditions.checkNotNull(transactionMetadataTasks, "transactionMetadataTasks");
+        Preconditions.checkNotNull(executor, "executor");
+        Preconditions.checkArgument(maxTxnTimeoutMillis > 0, "maxTxnTimeoutMillis should be a positive number");
+
+        this.streamMetadataStore = streamMetadataStore;
+        this.transactionMetadataTasks = transactionMetadataTasks;
+        this.maxTxnTimeoutMillis = maxTxnTimeoutMillis;
+        this.executor = executor;
+    }
+
+    public boolean isReady() {
+        return transactionMetadataTasks.isReady();
+    }
+
+    public void awaitInitialization() throws InterruptedException {
+        transactionMetadataTasks.awaitInitialization();
+    }
+
+    public CompletableFuture<Void> sweepFailedHosts(Supplier<Set<String>> activeHosts) {
+        if (!isReady()) {
+            return FutureHelpers.failedFuture(new IllegalStateException(getClass().getName() + " not yet ready"));
+        }
+        return streamMetadataStore.listHostsOwningTxn().thenComposeAsync(index -> {
+            index.removeAll(activeHosts.get());
+            log.info("Failed hosts {} have orphaned tasks", index);
+            return FutureHelpers.allOf(index.stream().map(this::sweepOrphanedTxns).collect(Collectors.toList()));
+        }, executor);
+    }
+
+    public CompletableFuture<Void> sweepOrphanedTxns(String failedHost) {
+        if (!isReady()) {
+            return FutureHelpers.failedFuture(new IllegalStateException(getClass().getName() + " not yet ready"));
+        }
+        log.info("Host={}, sweeping orphaned transactions", failedHost);
+        return FutureHelpers.delayedFuture(Duration.ofMillis(maxTxnTimeoutMillis), executor)
+                .thenComposeAsync(ignore ->
+                        FutureHelpers.doWhileLoop(() -> failOverTxns(failedHost), x -> x != null, executor), executor)
+                .whenCompleteAsync((v, e) ->
+                        log.info("Host={}, sweeping orphaned transactions complete", failedHost), executor);
+    }
+
+    private CompletableFuture<Result> failOverTxns(String failedHost) {
+        return streamMetadataStore.getRandomTxnFromIndex(failedHost).thenComposeAsync(resourceOpt -> {
+            if (resourceOpt.isPresent()) {
+                TxnResource resource = resourceOpt.get();
+                // Get the txn's status
+                // If it is aborting or committing send an abortEvent or commitEvent to respective streams
```

Comment:
`If it is aborting or committing, then send an abortEvent or commitEvent to respective streams`